src/interpret.jl

```diff
@@ -1,30 +1,47 @@
 using TimeZones.TZData: MIN_OFFSET, MAX_OFFSET
 
-# TimeZone concepts used to disambiguate context of DateTimes
-# abstract type UTC <: TimeZone end # Defined in Dates
-abstract type Local <: TimeZone end
+# Compare a local instant to a UTC transition instant by using the offset to make them both
+# into local time. We could just as easily convert both of them into UTC time.
+lt_local(local_dt::DateTime, t::Transition) = isless(local_dt, t.utc_datetime + t.zone.offset)
+lt_local(t::Transition, local_dt::DateTime) = isless(t.utc_datetime + t.zone.offset, local_dt)
+
+lt_utc(utc_dt::DateTime, t::Transition) = isless(utc_dt, t.utc_datetime)
+lt_utc(t::Transition, utc_dt::DateTime) = isless(t.utc_datetime, utc_dt)
 
 function transition_range(local_dt::DateTime, tz::VariableTimeZone, ::Type{Local})
-    transitions = tz.transitions
-
-    # Determine the earliest and latest possible UTC DateTime
-    # that this local DateTime could be.
-    # TODO: Alternatively we should only look at the range of offsets available within
-    # this TimeZone.
-    earliest = local_dt + MIN_OFFSET
-    latest = local_dt + MAX_OFFSET
-
-    # Determine the earliest transition the local DateTime could
-    # occur within.
-    start = searchsortedlast(
-        transitions, earliest,
-        by=el -> isa(el, Transition) ? el.utc_datetime : el,
-    )
-    start = max(start, 1)
-    finish = length(transitions)
-    for i in start:finish
-        if transitions[i].utc_datetime > latest
-            finish = i - 1
+    # To understand the logic in this function some background on transitions is needed:
+    #
+    # A transition (`t[i]`) is applicable to a given UTC instant that occurs on or after the
+    # transition start (`t[i].utc_datetime`). The transition (`t[i]`) ends at the start of
+    # the next transition in the list (`t[i + 1].utc_datetime`).
+    #
+    # Any UTC instant that occurs prior to the first transition (`t[1].utc_datetime`) has no
+    # associated transitions. Any UTC instant that occurs on or after the last transition
+    # (`t[end].utc_datetime`) is associated, at a minimum, with the last transition.
+
+    # Determine the latest transition that applies to `local_dt`. If the `local_dt`
+    # preceeds all transitions `finish` will be zero and produce the empty range `1:0`.
+    finish = searchsortedlast(tz.transitions, local_dt, lt=lt_local)
+
+    # Usually we'll begin by having `start` be larger than `finish` to create an empty
+    # range by default. In the scenario where last transition applies to the `local_dt` we
+    # can avoid a bounds by setting `start = finish`.
+    start = finish < length(tz.transitions) ? finish + 1 : finish
+
+    # To determine the first transition that applies to the `local_dt` we will work
+    # backwards. Typically, this loop will only use single iteration as multiple iterations
+    # only occur when local times are ambiguous.
+    @inbounds for i in (start - 1):-1:1
```

Comment:
Does this `@inbounds` make a difference?

Reply:
No, it's definitely inbounds though so it shouldn't hurt

Reply:
To elaborate further this loop typically sees a single iteration and two at most in real-world scenarios. I'm pretty sure there aren't enough in-bounds checks to actually make a difference to performance.

Reply:
> it's definitely inbounds though so it shouldn't hurt

with at-inbounds I worry when there isn't an explicit check since a bug in e.g. `searchsortedlast` could then cause Julia to crash

Reply:
If `start` was <= 0 then the range would be empty and we would never enter the loop. The only case that could happen is if `searchsortedlast` returned a value  > `length(tz.transactions)`. We could guarantee that by having:
```julia
start = finish < length(tz.transitions) ? finish + 1 : length(tz.transitions)
```
As I said it doesn't really help performance so I can also leave it out.

Reply:
Might as well leave it out then

Reply:
I double checked this was actually making an impact by defining the following function and running with n = 10000
```julia
f(dt, tz, n) = [TimeZones.transition_range(dt, tz, Local) for _ in 1:n][end]
```
Without `@inbounds`:
```julia
julia> @benchmark f($(DateTime(2015,11,1,1)), $(tz"America/Winnipeg"), 10000)
BenchmarkTools.Trial:
  memory estimate:  156.33 KiB
  allocs estimate:  2
  --------------
  minimum time:     237.635 μs (0.00% GC)
  median time:      320.934 μs (0.00% GC)
  mean time:        383.957 μs (4.50% GC)
  maximum time:     7.403 ms (93.06% GC)
  --------------
  samples:          10000
  evals/sample:     1
```
With `@inbounds`:
```julia
julia> @benchmark f($(DateTime(2015,11,1,1)), $(tz"America/Winnipeg"), 10000)
BenchmarkTools.Trial:
  memory estimate:  156.33 KiB
  allocs estimate:  2
  --------------
  minimum time:     221.501 μs (0.00% GC)
  median time:      298.603 μs (0.00% GC)
  mean time:        360.896 μs (4.79% GC)
  maximum time:     7.964 ms (91.46% GC)
  --------------
  samples:          10000
  evals/sample:     1
```
There does seem to be a noticeable difference when constructing a lot of `ZonedDateTime`s so I'll implement the `length` suggestion I had instead.